lib/mayaUsd/ufe/StagesSubject.cpp

```diff
@@ -316,15 +318,21 @@ void StagesSubject::stageChanged(UsdNotice::ObjectsChanged const& notice, UsdSta
 
 void StagesSubject::onStageSet(const MayaUsdProxyStageSetNotice& notice)
 {
-	// We should have no listerners and stage map is dirty.
-	TF_VERIFY(g_StageMap.isDirty());
-	TF_VERIFY(fStageListeners.empty());
-
-	StagesSubject::Ptr me(this);
-	for (auto stage : ProxyShapeHandler::getAllStages())
+	// Handle re-entrant onStageSet
+	if (!stageSetGuardCount.load())
```

Comment:
You will have atomic load, but the store on line 324 is not guaranteed to happen without any other thread reading the current value of stageSetGuardCount. Here is how you can make it (and then remove the line 324)
```suggestion
	bool expectedState = false;
	if (stageSetGuardCount.compare_exchange_strong(expectedState,true))
```